### PR TITLE
Allow to configure cache pool for Spanner client

### PR DESCRIPTION
**Pull request by gabrielfs7**
- Allow to configure cache pool for Spanner client
- Allow to Session pool Spanner Driver
- Allow inject all SpannerClient configurations through driver options

https://oat-sa.atlassian.net/browse/NEX-1829

I increased the coverage to 100% integration + unit tests:

Screenshot 2020-09-29 at 10 44 58

### CODE REVIEW

File: src/SpannerClient/SpannerClientFactory.php [hunk 1]
```diff
@@ -33,51 +35,72 @@ class SpannerClientFactory
...
+        return new CacheSessionPool(
+            new SysVCacheItemPool(['proj' => 'B']),
             [
                 'lock' => new SemaphoreLock(65535),
                 'minSessions' => self::SESSIONS_MIN,
                 'maxSessions' => self::SESSIONS_MAX,
             ]
         );
+    }
+
```
Comment (siwane) [inline src/SpannerClient/SpannerClientFactory.php:91]: Could u have exception doc here please?

File: src/SpannerClient/SpannerClientFactory.php [hunk 2]
```diff
@@ -33,51 +35,72 @@ class SpannerClientFactory
...
-        }
-
-        $keyFile = json_decode(file_get_contents($keyFileName), true);
-
-        return new SpannerClient(['keyFile' => $keyFile, 'authCache' => $authCache]);
+        return new SpannerClient(
+            array_merge(
+                [
+                    'keyFile' => $this->getKeyFileParsedContent(),
+                    'authCache' => $this->authCache ?? new SysVCacheItemPool()
```
Comment (siwane) [inline src/SpannerClient/SpannerClientFactory.php]: Could u move default dependencies to constructor please? (SysVCacheItemPool and configuration)

**Review by siwane — CHANGES_REQUESTED** [review]
What do u think about removing deprecated code, and move to a major version? 

- [x] New code is covered by tests (if applicable)
- [x] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [x] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful

File: src/SpannerClient/SpannerClientFactory.php (hunk 1 above)
Comment (gabrielfs7) [reply src/SpannerClient/SpannerClientFactory.php:91]: Done, thanks!

**Review by mike-ionut-mihai-sandu-tao — APPROVED** [review]
lgtm

File: src/SpannerClient/SpannerClientFactory.php [hunk 3]
```diff
@@ -33,51 +35,74 @@ class SpannerClientFactory
...
+        );
     }
 
     /**
-     * Creates a session pool to allow multiple sessions to share the auth cache.
+     * @TODO We must remove this method in the next major version
      *
-     * @return CacheSessionPool
-     * @throws \Exception
+     * @deprecated Please DO NOT use this method. The client of this library should define it if needed
```
Comment (boajer) [inline src/SpannerClient/SpannerClientFactory.php:78]: What prevents from removing deprecated  code from current version?

Comment (gabrielfs7) [reply src/SpannerClient/SpannerClientFactory.php:78]: We already received feedback from management that we are increasing major version in an unplanned way. We should deprecate code, plan a major version and do it. So I would prefer go this way. It is also a good habit, specially for extensions (which is not the case now, I know).

**Review by boajer — APPROVED** [review]

**Review by siwane — APPROVED** [review]
LGTM

**Pull request merged**